src/java/org/apache/cassandra/io/tries/IncrementalDeepTrieWriterPageAware.java

```diff
@@ -0,0 +1,405 @@
+        {
+            int sz = 0;
+            for (Node<VALUE> child : node.children)
+            {
+                if (depth < maxRecursionDepth)
+                    sz += recalcTotalSizeRecursiveOnStack(child, nodePosition + sz, depth + 1);
+                else
+                    sz += recalcTotalSizeRecursiveOnHeap(child, nodePosition + sz);
+            }
+            node.branchSize = sz;
+        }
+
+        // The sizing below will use the branch size calculated above. Since that can change on out-of-page in branch,
+        // we need to recalculate the size if either flag is set.
+        if (node.hasOutOfPageChildren || node.hasOutOfPageInBranch)
+            node.nodeSize = serializer.sizeofNode(node, nodePosition + node.branchSize);
+
+        return node.branchSize + node.nodeSize;
+    }
+
+    private int recalcTotalSizeRecursiveOnHeap(Node<VALUE> node, long nodePosition) throws IOException
+    {
+        if (node.hasOutOfPageInBranch)
+            new RecalcTotalSizeRecursion(node, null, nodePosition).process();
+
+        if (node.hasOutOfPageChildren || node.hasOutOfPageInBranch)
+            node.nodeSize = serializer.sizeofNode(node, nodePosition + node.branchSize);
+
+        return node.branchSize + node.nodeSize;
+    }
+
+    class RecalcTotalSizeRecursion extends Recursion<Node<VALUE>>
+    {
+        final long nodePosition;
+        int sz;
+
+        RecalcTotalSizeRecursion(Node<VALUE> node, Recursion<Node<VALUE>> parent, long nodePosition)
+        {
+            super(node, node.children.iterator(), parent);
+            sz = 0;
+            this.nodePosition = nodePosition;
+        }
+
+        @Override
+        Recursion<Node<VALUE>> makeChild(Node<VALUE> child)
+        {
+            if (child.hasOutOfPageInBranch)
+                return new RecalcTotalSizeRecursion(child, this, nodePosition + sz);
+            else
+                return null;
+        }
+
+        @Override
+        void complete()
+        {
+            node.branchSize = sz;
+        }
+
+        @Override
+        void completeChild(Node<VALUE> child)
+        {
+            // This will be called for nodes that were recursively processed as well as the ones that weren't.
+
+            // The sizing below will use the branch size calculated above. Since that can change on out-of-page in branch,
+            // we need to recalculate the size if either flag is set.
+            if (child.hasOutOfPageChildren || child.hasOutOfPageInBranch)
+            {
+                long childPosition = this.nodePosition + sz;
+                child.nodeSize = serializer.sizeofNode(child, childPosition + child.branchSize);
+            }
+
+            sz += child.branchSize + child.nodeSize;
+        }
+    }
+
+    @Override
+    protected long write(Node<VALUE> node) throws IOException
+    {
+        return writeRecursiveOnStack(node, 0);
+    }
+
+    private long writeRecursiveOnStack(Node<VALUE> node, int depth) throws IOException
```

Comment:
nit: Seems like the duplication is intended, so maybe throw a `@SuppressWarnings("DuplicatedCode")` on there.